R/reporter-junit.R

```diff
@@ -0,0 +1,153 @@
+    tests    = NULL,
+    root     = NULL,
+    suite    = NULL,
+    suite_time = NULL,
+
+    initialize = function(file = "report.xml") {
+      super$initialize()
+      self$file    <- file
+    },
+
+    elapsed_time = function() {
+      time <- round((proc.time() - self$timer)[["elapsed"]], 2)
+      self$timer  <- proc.time()
+      time
+    },
+
+    reset_suite = function () {
+      self$errors   <- 0
+      self$failures <- 0
+      self$skipped  <- 0
+      self$tests    <- 0
+      self$suite_time <- 0
+    },
+
+    start_reporter = function() {
+      self$timer <- proc.time()
+      self$doc   <- xml2::xml_new_document()
+      self$root  <- xml2::xml_add_child(self$doc, 'testsuites')
+      self$reset_suite()
+    },
+
+    start_context = function(context) {
+      self$cat(context, ": ")
+
+      self$suite <- xml2::xml_add_child(self$root, "testsuite")
+      add_attrs(self$suite,
+        name      = context,
+        timestamp = toString(Sys.time()),
+        hostname  = Sys.info()[["nodename"]]
+      )
+    },
+
+    end_context = function(context) {
+      self$cat("\n")
+
+      add_attrs(self$suite,
+        tests = self$tests, skipped = self$skipped,
+        failures = self$failures, errors = self$errors,
+        time = self$suite_time
+      )
+
+      self$reset_suite()
+    },
+
+    add_result = function(context, test, result) {
+      self$tests <- self$tests + 1
+
+      time <- self$elapsed_time()
+      self$suite_time <- self$suite_time + time
+
+      # XML node for test case
+      testcase <- xml2::xml_add_child(self$suite, "testcase",
+       time = time,
+       classname = paste0(classnameOK(context), '.', classnameOK(name))
+      )
+      name <- test %||% "(unnamed)"
+
+      # message - if failure or error
+      message <- if (is.null(result$call)) "(unexpected)" else format(result$call)[1]
+
+      if (!is.null(result$srcref)) {
+        location <- paste0('@', attr(result$srcref, 'srcfile')$filename, '#', result$srcref[1])
+        message  <- paste(as.character(result), location)
+      }
+
+      # add child XML node if not success
+      if (expectation_error(result)) {
+        self$cat_tight(single_letter_summary(result))
+        add_broken(testcase, 'error', message)
+        self$errors <- self$errors + 1
+      } else
+      if (expectation_failure(result)) {
+        self$cat_tight(single_letter_summary(result))
+        add_broken(testcase, 'failure', message)
+        self$failures <- self$failures + 1
+      } else
+      if (expectation_skip(result)) {
+        self$cat_tight(colourise("S", "success"))
+        xml2::xml_add_child(testcase, "skipped")
+        self$skipped <- self$skipped + 1
+      } else
+      {
+        self$cat_tight(colourise(".", "success"))
+      }
+    },
+
+    end_reporter = function() {
+      self$cat("\n")
+
+      # this causes a segfault write_xml(xmlDoc, self$file, format = )
+      cat(toString(self$doc), file = self$file)
```

Comment:
What if I'd like to output XML to a file? Do you expect people to `capture.output()` ?